Tracy profiler integration

divyaprakash-iitd opened the pull request:
## Proposed Changes
Integrated the Tracy profiler into the SU2 codebase with Meson support. Added build options for enabling profiling, instrumented the function Predict_MLP in SU2_CFD/src/fluid/CDataDrivenFluid.cpp with Tracy macros for demonstration of performance monitoring, and updated the configuration summary for clarity. These changes allow real-time, low-overhead performance analysis via the Tracy GUI.
 ~~Details of these changes are documented in my blog post: https://dpcfd.com/posts/2025/07/gsoc-week-3/~~
Details of these changes are here: https://github.com/divyaprakash-iitd/su2code.github.io/blob/develop/_docs_v7/Tracy-Integration.md


## Related Work
This PR does not resolve any open issues but contributes to profiling. This effort aligns with ongoing GSoC work.


## PR Checklist
- [x] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with --warnlevel=3 when using meson).
- [x] My contribution is commented and consistent with SU2 style (https://su2code.github.io/docs_v7/Style-Guide/).
- [x] I used the pre-commit hook to prevent dirty commits and used `pre-commit run --all` to format old commits.
- [x] I have added a test case that demonstrates my contribution, if necessary.
- [x] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp), if necessary.

bigfooted commented:
Hi Nice to have this, can you also add your blog post to the tutorial page on su2code.github.io, maybe under Developer Docs?

divyaprakash-iitd commented:
> Hi Nice to have this, can you also add your blog post to the tutorial page on su2code.github.io, maybe under Developer Docs?

I have added a new file, Tracy-Integration.md, to the _docs_v7 directory in the develop branch of the su2code/su2code.github.io repository and submitted a pull request. Is this the correct location for new documentation?

pcarruscag reviewed (commented):
Does this only profile the functions you annotate? Do you still see them as fraction of total time even if you don't annotate the main function of the code?

To integrate this type of tools that require intrusive changes to be used (i.e. annotations) we need to wrap them.

Similar to what we do for AD, OpenMP, etc., if tracy is included then you define some macros for the functionality that we want to expose (e.g. ZoneScopedN) and you turn on the include of tracy (it should not be included explicitly in any other files).
Have a look at omp_structure.hpp and follow the same pattern.

divyaprakash-iitd commented:
> Does this only profile the functions you annotate? Do you still see them as fraction of total time even if you don't annotate the main function of the code?
> 
> To integrate this type of tools that require intrusive changes to be used (i.e. annotations) we need to wrap them.
> 
> Similar to what we do for AD, OpenMP, etc., if tracy is included then you define some macros for the functionality that we want to expose (e.g. ZoneScopedN) and you turn on the include of tracy (it should not be included explicitly in any other files). Have a look at omp_structure.hpp and follow the same pattern.

Yes, it profiles only the annotated functions, and the functions still appear as a fraction of the total time without annotating the main function.
Thank you for your suggestions. I have looked at "omp_structure.hpp" and implemented a similar "tracy_structure.hpp". I have added documentation on using it to a separate [PR](https://github.com/su2code/su2code.github.io/pull/180). I have verified it by testing on my system. Please let me know if I should make any changes.

rois1995 commented on meson.build:
```diff
@@ -6,7 +6,8 @@ project('SU2', 'c', 'cpp',
                           'warning_level=0',
                           'c_std=c99',
-                          'cpp_std=c++17'])
+                          'cpp_std=c++11',
```
Why the change of c++ from 17 to 11? Is it required for Tracy?  From the documentation I think that c++11 is the lowest version possible. I think c++17 should remain the default and, if necessary, the c++11 will be set when running meson if tracy is on.

pcarruscag commented on Common/include/tracy_structure.hpp:
```diff
@@ -0,0 +1,42 @@
+
+#ifdef TRACY_ENABLE
+#  include "tracy/Tracy.hpp"
+#  define SU2_ZONE_SCOPED ZoneScoped
+#  define SU2_ZONE_SCOPED_N(name) ZoneScopedN(name)
```
Is the report of the profiling printed automatically? Or do we need to call a function somewhere?

divyaprakash-iitd replied:
Thank you! :smiley:

divyaprakash-iitd commented on meson_options.txt:
```diff
@@ -29,3 +29,7 @@ option('opdi-shared-read-opt', type : 'boolean', value : true, description : 'Op
+option('tracy_enable', 
+       type: 'boolean', 
+       value: false, 
+       description: 'Enable Tracy profiling support')
```
I have made the formatting consistent now

pcarruscag commented on meson.build:
```diff
@@ -7,7 +7,8 @@ project('SU2', 'c', 'cpp',
                           'b_ndebug=true',
                           'c_std=c99',
-                          'cpp_std=c++17'])
+                          'cpp_std=c++17',
+                          'enable-tracy=false'])
```
```suggestion
                          'cpp_std=c++17'])
```
